### PR TITLE
[CIRCT] [fud] Add MLIR backend to `fud`.

### LINKED ISSUE #648: [CIRCT] Add FUD support for CIRCT backend.
Neat; this is super snazzy. Maybe not for this PR, but perhaps Fud would like to learn `--to mlir` or `--to circt` for convenience as well?

_Originally posted by @sampsyo in https://github.com/cucapra/calyx/issues/638#issuecomment-908274048_

### LINKED ISSUE #666: [CIRCT] Unintended output.
https://github.com/cucapra/calyx/blob/9cbe07a2f34198410ca51872f7efc5b13a7d4d02/tests/backend/circt/attributes.expect#L1-L5

### PR DESCRIPTION
**Pull request by cgyurgyik**
- I've renamed CIRCT to MLIR, to avoid confusion.
- Add `fud` support for MLIR output. Currently defaults to using `lower guards` and `well-formed`.
- Closes #666.
- Closes #648.
- Nit: place `interpreter` in alphabetical order in `fud`'s stages.

Example usage: 

```
fud e tests/backend/circt/attributes.futil --to mlir
```

### CODE REVIEW

**Comment by rachitnigam** [comment]
Argh, I fixed #666 on master before seeing this PR.

**Comment by rachitnigam** [comment]
> I've renamed CIRCT to MLIR, to avoid confusion.

Confusion about? Also, if you want to do this then rename the backend file and the test directory to `mlir` as well.

File: tests/backend/circt/with-guards.expect [hunk 1]
```diff
@@ -24,29 +25,24 @@ calyx.component @main(%go: i1 {go=1}, %clk: i1 {clk=1}, %reset: i1 {reset=1}) ->
...
     }
     calyx.group @Group2 {
       calyx.assign %or0.left = %c1.out : i1
       calyx.assign %or0.right = %c0.done : i1
       calyx.assign %c1.in = %c1.out, %or0.out ? : i1
-      calyx.assign %or1.left = %c1.out : i1
-      calyx.assign %or1.right = %c0.done : i1
-      calyx.assign %and0.left = %c1.out : i1
-      calyx.assign %and0.right = %or1.out : i1
-      calyx.group_done %c1.done, %and0.out ? : i1
```
Comment (rachitnigam) [inline tests/backend/circt/with-guards.expect:49]: Why did these lines get deleted?

File: tests/backend/circt/with-guards.futil [hunk 2]
```diff
@@ -30,14 +30,15 @@ component main(@go go: 1, @clk clk: 1, @reset reset: 1) -> (@done done: 1) {
...
       a0.left = m0.read_data;
-      Group1[go] = 1'd0;
+      a0.right = 32'd1;
       c0.in = c0.out;
       Group1[done] = c0.done;
     }
     group Group2 {
       c1.in = (c1.out | c0.done) ? c1.out;
-      Group2[done] = (c1.out & 1'd1 & (c1.out | c0.done)) ? c1.done;
+      Group2[done] = (c1.out & 1'd1 & (c1.out | 1'd1)) ? c1.done;
```
Comment (rachitnigam) [inline tests/backend/circt/with-guards.futil:41]: This change will cause the Calyx to parse the guard as `c1.out` and not actually let us test the `lower-guards` is working properly. I recommend constructing a complex guard here.

**Comment by cgyurgyik** [comment]
> > I've renamed CIRCT to MLIR, to avoid confusion.
> 
> Confusion about? Also, if you want to do this then rename the backend file and the test directory to `mlir` as well.

CIRCT is a project, not a backend. The IR that we are outputting is MLIR. Done.

File: tests/backend/circt/with-guards.futil (hunk 2 above)
Comment (cgyurgyik) [reply tests/backend/circt/with-guards.futil]: Ah good catch. Fixed.

File: tests/backend/circt/with-guards.expect (hunk 1 above)
Comment (cgyurgyik) [reply tests/backend/circt/with-guards.expect:49]: Because below.

**Comment by rachitnigam** [comment]
The fmt checks was failing previously. When fixed, merge.

**Comment by cgyurgyik** [comment]
> The fmt checks was failing previously. When fixed, merge.

Done!

**Comment by rachitnigam** [comment]
Looks like fmt is still failing

**Pull request merged**